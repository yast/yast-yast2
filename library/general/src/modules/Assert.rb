# encoding: utf-8

# ***************************************************************************
#
# Copyright (c) 2002 - 2012 Novell, Inc.
# All Rights Reserved.
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of version 2 of the GNU General Public License as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, contact Novell, Inc.
#
# To contact Novell about this file by physical or electronic mail,
# you may find current contact information at www.novell.com
#
# ***************************************************************************
require "yast"

module Yast
  # Provides assertions for old yast testsuite
  # @deprecated use rspec tests instead
  class AssertClass < Module
    # @param [Object] expected expected value of test
    # @param [Object] actual   actual value of test
    # @param [String] fail_message will be logged if test fails
    # @return whether test succeeds
    def EqualMsg(expected, actual, fail_message)
      expected = deep_copy(expected)
      actual = deep_copy(actual)
      return true if expected == actual

      Builtins.y2error("%1", fail_message)
      false
    end

    # @param [Object] expected expected value of test
    # @param [Object] actual   actual value of test
    # @return whether test succeeds
    def Equal(expected, actual)
      expected = deep_copy(expected)
      actual = deep_copy(actual)
      fail_message = Builtins.sformat(
        "assertion failure, expected '%1', got '%2'",
        expected,
        actual
      )
      EqualMsg(expected, actual, fail_message)
    end

    publish function: :EqualMsg, type: "boolean (any, any, string)"
    publish function: :Equal, type: "boolean (any, any)"
  end

  Assert = AssertClass.new
end
