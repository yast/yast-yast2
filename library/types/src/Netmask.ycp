/**
 * File:	modules/Netmask.ycp
 * Module:	yast2
 * Summary:	Netmask manipulation routines
 * Authors:	Michal Svec <msvec@suse.cz>
 * Flags:	Stable
 *
 * $Id$
 */

{

module "Netmask";
textdomain "base";

global string ValidChars = "0123456789.";
global string ValidChars4 = "0123456789.";
global string ValidChars6 = "0123456789";

global define boolean CheckPrefix4(string prefix){
    if(prefix == nil || prefix == "") return false;
    /* <0,32> */
    if(regexpmatch(prefix, "^[0-9]+$")) {
	integer nm = tointeger(prefix);
	return nm >= 0 && nm <= 32;
    } else
	return false;
}

/**
 * Check the IPv4 netmask
 * Note that 0.0.0.0 is not a correct netmask.
 * @param netmask network mask
 * @return true if correct
 */
global define boolean Check4(string netmask) ``{

    if(netmask == nil || netmask == "") return false;

    /* <0,32> */
    if(regexpmatch(netmask, "^[0-9]+$")) {
	integer nm = tointeger(netmask);
	return nm >= 0 && nm <= 32;
    }

    /* 255.255.240.0 */
    string s1 = "(128|192|224|240|248|252|254|255)";
    string nm = "^(" + s1 + ".0.0.0|" + "255." + s1 + ".0.0|"
	+ "255.255." + s1 + ".0|" + "255.255.255." + s1 + ")$";
    return regexpmatch(netmask, nm);
}

/**
 * Check the IPv6 netmask
 * @param netmask network mask
 * @return true if correct
 */
global define boolean Check6(string netmask) ``{

    if(netmask == nil || netmask == "") return false;

    /* <0,256> */
    if(!regexpmatch(netmask, "^[0-9]+$")) return false;
    integer nm = tointeger(netmask);
    return nm >= 0 && nm <= 256;
}

/**
 * Check the netmask
 * @param netmask network mask
 * @return true if correct
 */
global define boolean Check(string netmask) ``{
    return Check4(netmask) || Check6(netmask);
}

/**
 * Convert netmask in bits form (20) to IPv4 netmask string (255.255.240.0)
 *
 * @param bits  number of bits in netmask
 * @return      netmask string or empty string in case of invalid bits (e.g.
 *              when prefix is incompatible with IPv4)
 */
global define string FromBits(integer bits) ``{
    if( ( bits < 0) || (bits > 32) )
    {
        return "";
    }

    integer b = bits / 8;
    integer d = bits % 8;

    map l = $[ 1:"255.", 2:"255.255.", 3:"255.255.255.", 4:"255.255.255.255" ];
    map r = $[ 3:"0", 2:"0.0", 1:"0.0.0", 0:"0.0.0.0" ];
    map m = $[
	1:"128", 2:"192", 3:"224", 4:"240",
	5:"248", 6:"252", 7:"254", 8:"255",
    ];

    return l[b]:"" + (d!=0? m[d]:""+(b!=3?".":"") : "") + r[d==0?b:(b+1)]:"";
}

/**
 * Convert IPv4 netmask as string (255.255.240.0) to bits form (20)
 * @param netmask netmask as string
 * @return number of bits in netmask; 32 for empty string
 */
global define integer ToBits(string netmask) ``{
    if (netmask == "")
    {
	return 32;
    }
    integer bits = 0;
    map m = $[
	"128":1, "192":2, "224":3, "240":4,
	"248":5, "252":6, "254":7, "255":8,
    ];
    maplist(string i, splitstring(netmask, "."), ``{ bits = bits + m[i]:0; });
    return bits;
}

/* EOF */
}
