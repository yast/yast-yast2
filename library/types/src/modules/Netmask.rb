# encoding: utf-8

# ***************************************************************************
#
# Copyright (c) 2002 - 2012 Novell, Inc.
# All Rights Reserved.
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of version 2 of the GNU General Public License as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, contact Novell, Inc.
#
# To contact Novell about this file by physical or electronic mail,
# you may find current contact information at www.novell.com
#
# ***************************************************************************
# File:	modules/Netmask.ycp
# Module:	yast2
# Summary:	Netmask manipulation routines
# Authors:	Michal Svec <msvec@suse.cz>
# Flags:	Stable
#
# $Id$
require "yast"

module Yast
  class NetmaskClass < Module
    def main
      textdomain "base"

      @ValidChars = "0123456789."
      @ValidChars4 = "0123456789."
      @ValidChars6 = "0123456789"
    end

    def CheckPrefix4(prefix)
      return false if prefix == nil || prefix == ""
      # <0,32>
      if Builtins.regexpmatch(prefix, "^[0-9]+$")
        nm = Builtins.tointeger(prefix)
        return Ops.greater_or_equal(nm, 0) && Ops.less_or_equal(nm, 32)
      else
        return false
      end
    end

    # Check the IPv4 netmask
    # Note that 0.0.0.0 is not a correct netmask.
    # @param [String] netmask network mask
    # @return true if correct
    def Check4(netmask)
      return false if netmask == nil || netmask == ""


      # 255.255.240.0
      s1 = "(128|192|224|240|248|252|254|255)"
      nm = Ops.add(
        Ops.add(
          Ops.add(
            Ops.add(
              Ops.add(
                Ops.add(
                  Ops.add(
                    Ops.add(
                      Ops.add(Ops.add(Ops.add("^(", s1), ".0.0.0|"), "255."),
                      s1
                    ),
                    ".0.0|"
                  ),
                  "255.255."
                ),
                s1
              ),
              ".0|"
            ),
            "255.255.255."
          ),
          s1
        ),
        ")$"
      )
      Builtins.regexpmatch(netmask, nm)
    end

    # Check the IPv6 netmask
    # @param [String] netmask network mask
    # @return true if correct
    def Check6(netmask)
      return false if netmask == nil || netmask == ""

      # <0,256>
      return false if !Builtins.regexpmatch(netmask, "^[0-9]+$")
      nm = Builtins.tointeger(netmask)
      Ops.greater_or_equal(nm, 0) && Ops.less_or_equal(nm, 256)
    end

    # Check the netmask
    # @param [String] netmask network mask
    # @return true if correct
    def Check(netmask)
      Check4(netmask) || Check6(netmask)
    end

    #
    # Convert netmask in bits form (20) to IPv4 netmask string (255.255.240.0) 
    #
    # @param bits  number of bits in netmask 
    # @return      netmask string or empty string in case of invalid bits (e.g. 
    #              when prefix is incompatible with IPv4)
    #
    def FromBits(bits)
      return "" unless bits.between?( 0, 32)

      b = Ops.divide(bits, 8)
      d = Ops.modulo(bits, 8)

      l = {
        1 => "255.",
        2 => "255.255.",
        3 => "255.255.255.",
        4 => "255.255.255.255"
      }
      r = { 3 => "0", 2 => "0.0", 1 => "0.0.0", 0 => "0.0.0.0" }
      m = {
        1 => "128",
        2 => "192",
        3 => "224",
        4 => "240",
        5 => "248",
        6 => "252",
        7 => "254",
        8 => "255"
      }

      Ops.add(
        Ops.add(
          Ops.get_string(l, b, ""),
          d != 0 ? Ops.add(Ops.get_string(m, d, ""), b != 3 ? "." : "") : ""
        ),
        Ops.get_string(r, d == 0 ? b : Ops.add(b, 1), "")
      )
    end

    # Convert IPv4 netmask as string (255.255.240.0) to bits form (20)
    # @param [String] netmask netmask as string
    # @return number of bits in netmask; 32 for empty string
    def ToBits(netmask)
      return 32 if netmask == ""
      bits = 0
      m = {
        "128" => 1,
        "192" => 2,
        "224" => 3,
        "240" => 4,
        "248" => 5,
        "252" => 6,
        "254" => 7,
        "255" => 8
      }
      Builtins.maplist(Builtins.splitstring(netmask, ".")) do |i|
        bits = Ops.add(bits, Ops.get_integer(m, i, 0))
      end
      bits
    end

    publish :variable => :ValidChars, :type => "string"
    publish :variable => :ValidChars4, :type => "string"
    publish :variable => :ValidChars6, :type => "string"
    publish :function => :CheckPrefix4, :type => "boolean (string)"
    publish :function => :Check4, :type => "boolean (string)"
    publish :function => :Check6, :type => "boolean (string)"
    publish :function => :Check, :type => "boolean (string)"
    publish :function => :FromBits, :type => "string (integer)"
    publish :function => :ToBits, :type => "integer (string)"
  end

  Netmask = NetmaskClass.new
  Netmask.main
end
