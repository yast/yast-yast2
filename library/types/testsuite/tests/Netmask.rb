# encoding: utf-8

# ***************************************************************************
#
# Copyright (c) 2002 - 2012 Novell, Inc.
# All Rights Reserved.
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of version 2 of the GNU General Public License as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, contact Novell, Inc.
#
# To contact Novell about this file by physical or electronic mail,
# you may find current contact information at www.novell.com
#
# ***************************************************************************
module Yast
  class NetmaskClient < Client
    def main
      Yast.include self, "testsuite.rb"
      Yast.import "Netmask"

      DUMP("Netmask::Check4")
      TEST(->() { Netmask.Check4("128.0.0.0") }, [], nil)
      TEST(->() { Netmask.Check4("192.0.0.0") }, [], nil)
      TEST(->() { Netmask.Check4("224.0.0.0") }, [], nil)
      TEST(->() { Netmask.Check4("240.0.0.0") }, [], nil)
      TEST(->() { Netmask.Check4("248.0.0.0") }, [], nil)
      TEST(->() { Netmask.Check4("252.0.0.0") }, [], nil)
      TEST(->() { Netmask.Check4("254.0.0.0") }, [], nil)
      TEST(->() { Netmask.Check4("255.0.0.0") }, [], nil)
      TEST(->() { Netmask.Check4("255.128.0.0") }, [], nil)
      TEST(->() { Netmask.Check4("255.192.0.0") }, [], nil)
      TEST(->() { Netmask.Check4("255.224.0.0") }, [], nil)
      TEST(->() { Netmask.Check4("255.240.0.0") }, [], nil)
      TEST(->() { Netmask.Check4("255.248.0.0") }, [], nil)
      TEST(->() { Netmask.Check4("255.252.0.0") }, [], nil)
      TEST(->() { Netmask.Check4("255.254.0.0") }, [], nil)
      TEST(->() { Netmask.Check4("255.255.0.0") }, [], nil)
      TEST(->() { Netmask.Check4("255.255.128.0") }, [], nil)
      TEST(->() { Netmask.Check4("255.255.192.0") }, [], nil)
      TEST(->() { Netmask.Check4("255.255.224.0") }, [], nil)
      TEST(->() { Netmask.Check4("255.255.240.0") }, [], nil)
      TEST(->() { Netmask.Check4("255.255.248.0") }, [], nil)
      TEST(->() { Netmask.Check4("255.255.252.0") }, [], nil)
      TEST(->() { Netmask.Check4("255.255.254.0") }, [], nil)
      TEST(->() { Netmask.Check4("255.255.255.0") }, [], nil)
      TEST(->() { Netmask.Check4("255.255.255.128") }, [], nil)
      TEST(->() { Netmask.Check4("255.255.255.192") }, [], nil)
      TEST(->() { Netmask.Check4("255.255.255.224") }, [], nil)
      TEST(->() { Netmask.Check4("255.255.255.240") }, [], nil)
      TEST(->() { Netmask.Check4("255.255.255.248") }, [], nil)
      TEST(->() { Netmask.Check4("255.255.255.252") }, [], nil)
      TEST(->() { Netmask.Check4("255.255.255.254") }, [], nil)
      TEST(->() { Netmask.Check4("255.255.255.255") }, [], nil)
      TEST(->() { Netmask.Check4("1.2.3.4") }, [], nil)
      TEST(->() { Netmask.Check4("1.A.3.4") }, [], nil)
      TEST(->() { Netmask.Check4("0.0.0.0") }, [], nil)

      TEST(->() { Netmask.CheckPrefix4(nil) }, [], nil)
      TEST(->() { Netmask.CheckPrefix4("") }, [], nil)
      TEST(->() { Netmask.CheckPrefix4("33") }, [], nil)
      TEST(->() { Netmask.CheckPrefix4("0") }, [], nil)
      TEST(->() { Netmask.CheckPrefix4("24") }, [], nil)
      TEST(->() { Netmask.CheckPrefix4("32") }, [], nil)

      DUMP("Netmask::Check6")
      TEST(->() { Netmask.Check6(nil) }, [], nil)
      TEST(->() { Netmask.Check6("") }, [], nil)
      TEST(->() { Netmask.Check6("345") }, [], nil)
      TEST(->() { Netmask.Check6("128") }, [], nil)

      @i = nil
      DUMP("Netmask::FromBits")
      @i = 32
      while Ops.greater_or_equal(@i, 0)
        TEST(->() { Netmask.FromBits(@i) }, [], nil)
        @i = Ops.subtract(@i, 1)
      end

      # this test relies on the previous one
      DUMP("Netmask::ToBits")
      @i = 32
      while Ops.greater_or_equal(@i, 0)
        TEST(->() { Netmask.ToBits(Netmask.FromBits(@i)) }, [], nil)
        @i = Ops.subtract(@i, 1)
      end
      TEST(->() { Netmask.ToBits("") }, [], nil)

      nil
    end
  end
end

Yast::NetmaskClient.new.main
