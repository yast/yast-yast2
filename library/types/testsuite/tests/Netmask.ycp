{

include "testsuite.ycp";
import "Netmask";
import "Assert";

DUMP("Netmask::Check4");
TEST(``(Netmask::Check4("128.0.0.0")), [], nil);
TEST(``(Netmask::Check4("192.0.0.0")), [], nil);
TEST(``(Netmask::Check4("224.0.0.0")), [], nil);
TEST(``(Netmask::Check4("240.0.0.0")), [], nil);
TEST(``(Netmask::Check4("248.0.0.0")), [], nil);
TEST(``(Netmask::Check4("252.0.0.0")), [], nil);
TEST(``(Netmask::Check4("254.0.0.0")), [], nil);
TEST(``(Netmask::Check4("255.0.0.0")), [], nil);
TEST(``(Netmask::Check4("255.128.0.0")), [], nil);
TEST(``(Netmask::Check4("255.192.0.0")), [], nil);
TEST(``(Netmask::Check4("255.224.0.0")), [], nil);
TEST(``(Netmask::Check4("255.240.0.0")), [], nil);
TEST(``(Netmask::Check4("255.248.0.0")), [], nil);
TEST(``(Netmask::Check4("255.252.0.0")), [], nil);
TEST(``(Netmask::Check4("255.254.0.0")), [], nil);
TEST(``(Netmask::Check4("255.255.0.0")), [], nil);
TEST(``(Netmask::Check4("255.255.128.0")), [], nil);
TEST(``(Netmask::Check4("255.255.192.0")), [], nil);
TEST(``(Netmask::Check4("255.255.224.0")), [], nil);
TEST(``(Netmask::Check4("255.255.240.0")), [], nil);
TEST(``(Netmask::Check4("255.255.248.0")), [], nil);
TEST(``(Netmask::Check4("255.255.252.0")), [], nil);
TEST(``(Netmask::Check4("255.255.254.0")), [], nil);
TEST(``(Netmask::Check4("255.255.255.0")), [], nil);
TEST(``(Netmask::Check4("255.255.255.128")), [], nil);
TEST(``(Netmask::Check4("255.255.255.192")), [], nil);
TEST(``(Netmask::Check4("255.255.255.224")), [], nil);
TEST(``(Netmask::Check4("255.255.255.240")), [], nil);
TEST(``(Netmask::Check4("255.255.255.248")), [], nil);
TEST(``(Netmask::Check4("255.255.255.252")), [], nil);
TEST(``(Netmask::Check4("255.255.255.254")), [], nil);
TEST(``(Netmask::Check4("255.255.255.255")), [], nil);
TEST(``(Netmask::Check4("1.2.3.4")), [], nil);
TEST(``(Netmask::Check4("1.A.3.4")), [], nil);
TEST(``(Netmask::Check4("0.0.0.0")), [], nil);

TEST(``(Netmask::Check4(nil)), [], nil);
TEST(``(Netmask::Check4("")), [], nil);
TEST(``(Netmask::Check4("33")), [], nil);
TEST(``(Netmask::Check4("0")), [], nil);
TEST(``(Netmask::Check4("24")), [], nil);
TEST(``(Netmask::Check4("32")), [], nil);

DUMP("Netmask::Check6");
TEST(``(Netmask::Check6(nil)), [], nil);
TEST(``(Netmask::Check6("")), [], nil);
TEST(``(Netmask::Check6("345")), [], nil);
TEST(``(Netmask::Check6("128")), [], nil);

integer i = nil;
DUMP("Netmask::FromBits");
i = 32;
while (i >= 0) {
    TEST(``(Netmask::FromBits (i)), [], nil);
    i = i - 1;
}

// tests that returns empty mask for IPv6 specific prefixes (>32, <=128)
// implementatio of FromBits is IPv4 specific
i = 128;
while( i > 32)
{
    Assert::Equal( "", Netmask::FromBits(i) );
    i = i - 1;
}

// this test relies on the previous one
DUMP("Netmask::ToBits");
i = 32;
while (i >= 0) {
    TEST(``(Netmask::ToBits (Netmask::FromBits (i))), [], nil);
    i = i - 1;
}
TEST(``(Netmask::ToBits ("")), [], nil);

}
