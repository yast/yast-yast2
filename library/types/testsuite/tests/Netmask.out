Dump	Netmask::Check4
Return	true
Return	true
Return	true
Return	true
Return	true
Return	true
Return	true
Return	true
Return	true
Return	true
Return	true
Return	true
Return	true
Return	true
Return	true
Return	true
Return	true
Return	true
Return	true
Return	true
Return	true
Return	true
Return	true
Return	true
Return	true
Return	true
Return	true
Return	true
Return	true
Return	true
Return	true
Return	true
Return	false
Return	false
Return	false
Return	false
Return	false
Return	false
Return	true
Return	true
Return	true
Return	false
Return	false
Return	false
Return	true
Return	true
Return	true
Dump	Netmask::Check6
Return	false
Return	false
Return	false
Return	true
Dump	Netmask::FromBits
Return	255.255.255.255
Return	255.255.255.254
Return	255.255.255.252
Return	255.255.255.248
Return	255.255.255.240
Return	255.255.255.224
Return	255.255.255.192
Return	255.255.255.128
Return	255.255.255.0
Return	255.255.254.0
Return	255.255.252.0
Return	255.255.248.0
Return	255.255.240.0
Return	255.255.224.0
Return	255.255.192.0
Return	255.255.128.0
Return	255.255.0.0
Return	255.254.0.0
Return	255.252.0.0
Return	255.248.0.0
Return	255.240.0.0
Return	255.224.0.0
Return	255.192.0.0
Return	255.128.0.0
Return	255.0.0.0
Return	254.0.0.0
Return	252.0.0.0
Return	248.0.0.0
Return	240.0.0.0
Return	224.0.0.0
Return	192.0.0.0
Return	128.0.0.0
Return	0.0.0.0
Dump	Netmask::ToBits
Return	32
Return	31
Return	30
Return	29
Return	28
Return	27
Return	26
Return	25
Return	24
Return	23
Return	22
Return	21
Return	20
Return	19
Return	18
Return	17
Return	16
Return	15
Return	14
Return	13
Return	12
Return	11
Return	10
Return	9
Return	8
Return	7
Return	6
Return	5
Return	4
Return	3
Return	2
Return	1
Return	0
Return	32
